### ISSUE TITLE
experiments on imbalanced data

### ISSUE THREAD

**Issue by tdhock** [mlr3-imbalanced/mlr3torchAUM#87]
https://github.com/tdhock/2023-res-baz-az/blob/main/data_Classif_unbalanced.tgz contains csv files for assigning images to balanced or imbalanced subsets.

the *MNIST data files are https://github.com/tdhock/cv-same-other-paper?tab=readme-ov-file#data-set-files

**Comment by weicaocw** [comment]
@AmanKashyap0807  Thank you. Very nice.

Is it possible to submit the benchmark code into the repo as PRs? @tdhock @AmanKashyap0807

**Comment by AmanKashyap0807** [comment]
Hi @weicaocw ,

I was about to raise pr for this and also created a branch named `benchmarking` . But then i just shared the benchmark here instead of creating pr.

**Comment by tdhock** [comment]
interesting thanks for sharing.
your results show pairwise squared hinge learns faster than logistic.
this is consistent with my previous results that AUM learns faster than logistic.
some of the AUC values are very close to 1, so difficult to see details of how close it gets.
can you please try using `mlr3torchAUM::MeasureClassifInvAUC$new()` to show details?

**Comment by AmanKashyap0807** [comment]
Hi @tdhock ,
Thanks for the feedback and suggestions.

Below is the benchmark using `mlr3torchAUM::MeasureClassifInvAUC$new()`, clearly showing the details and convergence as AUC approaches 1.

<img width="2700" height="2100" alt="Image" src="https://github.com/user-attachments/assets/601117fb-61d4-4bbb-9aae-3d0ab0927475" />

<details>
<summary><b>show code</b>
</summary>

```r
library(data.table)
library(mlr3)
library(mlr3pipelines)
library(mlr3torch)
library(mlr3resampling)
library(torch)
library(ggplot2)
library(nc)
library(R6)
source("R/pairwise_auc_loss.R")
source("R/squared_hinge_loss.R")
source("R/MeasureClassifROCAUM.R")

set.seed(42)
if (torch::torch_is_installed()) {
  torch::torch_manual_seed(42)
}

(SOAK <- mlr3resampling::ResamplingSameOtherSizesCV$new())
unb_files <- c("EMNIST.csv", "FashionMNIST.csv")
task_list <- list()
for (f in unb_files) {
  unb_csv <- file.path("benchmarking/data_Classif_unbalanced", f)
  data_csv <- file.path("benchmarking/data_Classif", f)
  if (!file.exists(unb_csv) || !file.exists(data_csv)) next
  MNIST_dt <- fread(data_csv)
  subset_dt <- fread(unb_csv)
  task_dt <- data.table(subset_dt, MNIST_dt)[, odd := factor(y %% 2)]
  feature_names <- grep("^[0-9]+$", names(task_dt), value = TRUE)
  subset_name <- "seed1_prop0.01"
  data_name <- gsub("[.]csv$", "", f)
  subset_vec <- task_dt[[subset_name]]
  task_id <- paste0(data_name, "_", subset_name)
  itask <- mlr3::TaskClassif$new(
    id = task_id,
    backend = task_dt[subset_vec != ""],
    target = "odd"
  )
  itask$col_roles$stratum <- "y"
  itask$col_roles$subset <- subset_name
  itask$col_roles$feature <- feature_names
  task_list[[task_id]] <- itask
}

cat("Successfully prepared tasks:\n")
print(names(task_list))

nn_bce_loss3 <- torch::nn_module(
  c("nn_bce_with_logits_loss3", "nn_loss"),
  initialize = function(weight = NULL, reduction = "mean", pos_weight = NULL) {
    self$loss <- torch::nn_bce_with_logits_loss(weight, reduction, pos_weight)
  },
  forward = function(input, target) {
    target_float <- target$reshape(-1)$to(dtype = torch::torch_float())
    if (torch::as_array(target_float$max() == 2)) target_float <- target_float - 1
    self$loss(input$reshape(-1), target_float)
  }
)

nn_pairwise_auc_sq_hinge <- torch::nn_module(
  c("nn_pairwise_auc_sq_hinge", "nn_loss"),
  initialize = function(margin = 1.0) {
    self$loss_fn <- nn_pairwise_auc_loss(surr_loss = "squared_hinge", margin = margin)
  },
  forward = function(input, target) {
    input_1d <- input$reshape(-1)
    target_1d <- target$reshape(-1)$to(dtype = torch::torch_long())
    target_01 <- target_1d
    if (torch::as_array(target_01$max() == 2L)) {
      target_01 <- target_01 - 1L
    }
    if (torch::as_array(target_01$sum() == 0L) || torch::as_array(target_01$sum() == target_01$numel())) {
      return(torch::torch_sum(input_1d * 0))
    }
    self$loss_fn$forward(input_1d, target_01)
  }
)

nn_sq_hinge_loglinear <- torch::nn_module(
  c("nn_sq_hinge_loglinear_mod", "nn_loss"),
  initialize = function(margin = 1.0) {
    self$loss_fn <- nn_squared_hinge_loss(margin = margin)
  },
  forward = function(input, target) {
    input_1d <- input$reshape(-1)
    target_1d <- target$reshape(-1)$to(dtype = torch::torch_long())
    target_01 <- target_1d
    if (torch::as_array(target_01$max() == 2L)) {
      target_01 <- target_01 - 1L
    }
    if (torch::as_array(target_01$sum() == 0L) || torch::as_array(target_01$sum() == target_01$numel())) {
      return(torch::torch_sum(input_1d * 0))
    }
    self$loss_fn$forward(input_1d, target_01)
  }
)

MAX_EPOCHS <- 300
measure_list <- mlr3::msrs(c("classif.auc", "classif.acc"))
measure_list$classif.invauc <- MeasureClassifInvAUC$new()
make_torch_learner <- function(id, loss_fun) {
  po_list <- list(
    mlr3pipelines::po("select", selector = mlr3pipelines::selector_type(c("numeric", "integer"))),
    mlr3torch::PipeOpTorchIngressNumeric$new(),
    mlr3torch::nn("linear", out_features = 1),
    mlr3pipelines::po("torch_loss", loss_fun),
    mlr3pipelines::po("torch_optimizer", mlr3torch::t_opt("adam", lr = 0.01)),
    mlr3pipelines::po("torch_callbacks", mlr3torch::t_clbk("history")),
    mlr3pipelines::po(
      "torch_model_classif",
      batch_size = 100000,
      measures_valid = measure_list,
      measures_train = measure_list,
      predict_type = "prob",
      device = "cuda",
      epochs = MAX_EPOCHS
    )
  )
  graph <- Reduce(mlr3pipelines::concat_graphs, po_list)
  glearner <- mlr3::as_learner(graph)
  mlr3::set_validate(glearner, validate = 0.5)
  glearner$id <- id
  glearner
}

learner_list <- list(
  logistic = make_torch_learner("logistic", nn_bce_loss3),
  pairwise_auc_sq_hinge = make_torch_learner("pairwise_auc_sq_hinge", nn_pairwise_auc_sq_hinge),
  sq_hinge_loglinear = make_torch_learner("sq_hinge_loglinear", nn_sq_hinge_loglinear)
)

for (learner_i in seq_along(learner_list)) {
  learner_list[[learner_i]]$predict_type <- "prob"
}

cat("\n Building Benchmark Grid \n")
bench_grid <- mlr3::benchmark_grid(task_list, learner_list, SOAK)
cat(" Executing Benchmark \n")
bench_result <- mlr3::benchmark(bench_grid, store_models = TRUE)
cat("\nExtracting History Logs & Plotting Curves\n")
cat("Calling mlr3resampling::score\n")
score_dt <- mlr3resampling::score(bench_result, mlr3::msr("classif.auc"))
cat("Score DT computed!\n")

extract_history <- function(lrn_obj) {
  if (inherits(lrn_obj, "GraphLearner")) {
    torch_mdl <- lrn_obj$model$torch_model_classif$model
  } else {
    torch_mdl <- lrn_obj$model
  }
  torch_mdl$callbacks$history
}

cat("Extracting history DT...\n")
history_dt <- score_dt[
  , extract_history(learner[[1]])
  , by = .(task_id, test.subset, train.subsets, test.fold, algorithm)
]
cat("History extracted!\n")

melt_history <- function(DT) {
  nc::capture_melt_single(
    DT,
    set = nc::alevels(train = "subtrain", valid = "validation"),
    ".classif.",
    measure = nc::alevels(acc = "accuracy_prop", ce = "error_prop", auc = "AUC", invauc = "InvAUC", "logloss")
  )
}

history_long <- melt_history(history_dt)[
  , train.subset := ifelse(train.subsets == "same", "balanced", "unbalanced")
]

query <- function(DT) {
  DT[test.subset == "balanced" & train.subsets != "all" & grepl("seed1", task_id)]
}

history_show <- query(
  history_long[test.fold == 1 & measure == "InvAUC" & train.subsets != "same"]
)[, let(
  Data = gsub("_.*", "", task_id)
)]

p <- ggplot(history_show) +
  theme_bw(base_size = 12) +
  geom_hline(yintercept = 0.5, color = "grey", linetype = "dashed") +
  geom_line(aes(epoch, value, color = set), linewidth = 0.8) +
  facet_grid(algorithm ~ Data, scales = "free_y", labeller = label_both) +
  scale_y_log10() +
  scale_color_manual(values = c("subtrain" = "#F8766D", "validation" = "#00BFC4")) +
  labs(
    x = "epoch",
    y = "InvAUC (unbalanced train set)",
    color = "set"
  ) +
  theme(
    legend.position = "right",
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill = "grey90"),
    strip.text = element_text(size = 11, face = "bold")
  )

output_plot_path <- "benchmarking/auc_epoch_curves.png"
ggsave(output_plot_path, plot = p, width = 9, height = 7, dpi = 300)
cat(sprintf("Generated and saved plot to %s\n", output_plot_path))
```
</details>

Please let me know for any changes and suggestions for further benchmarking.

**Comment by tdhock** [comment]
this is figure shows more details, thanks!
the result shows that the validation curve is minimized before 100 epochs, with the squared hinge losses.
and that the logistic loss takes over 300 epochs.
all algos use adam optimizer with lr=0.01 right?
this comparison seems fair enough for a preliminary demo, but to be even more fair, we could try different lr values.
I wonder if we increased lr, maybe logistic could get a min valid curve in a smaller number of epochs?

anyway, this is good enough for validating the current code.

**Comment by AmanKashyap0807** [comment]
Hi @tdhock , yes, all learners use Adam with lr=0.01. Happy to run a follow-up comparing different lr values if that would be useful, just let me know whether you'd like a new issue for that or if the current results are enough for now.
